File: Applications/Python/LinkedList/ReorderLinkedList.py
# 143. Reorder List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reorderList(self, head):
        """
        Do not return anything, modify head in-place instead.
        """
        if head is None or head.next is None:
            return

        result = head.next
        lst = []

        while(result != None):
            lst.append(result)
            result = result.next


        l = 0
        r = len(lst) - 1
        result = head
        while (l <= r):
            result.next = lst[r]
            result.next.next = lst[l]
            result = result.next.next

            r -= 1
            l += 1

        result.next = None

File: Applications/CSharp/LinkedList/ReorderLinkedList.cs
class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}

class Solution
{
    public void ReorderList(ListNode head)
    {
        if (head == null || head.next == null)
            return;

        var result = head.next;
        var list = new List<ListNode>();
        while (result != null)
        {
            list.Add(result);
            result = result.next;
        }

        var l = 0;
        var r = list.Count - 1;
        result = head;
        while (l <= r)
        {
            result.next = list[r--];
            result.next.next = list[l++];
            result = result.next.next;
        }

        if (l == r)
        {
            result.next = list[l];
        }
        result.next = null;
    }
}